src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1378,43 +1363,64 @@ private static bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol t
             Debug.Assert((object)type1 != null);
             Debug.Assert((object)type2 != null);
 
-            // https://github.com/dotnet/roslyn/issues/27961: If two types differ only by nullability and
-            // one has IsNullable unset and the other doesn't, which do we choose in inference?
-            // See NullableReferenceTypesTests.TypeInference_04.
-            var compareKind = includeNullability ?
-                TypeCompareKind.AllIgnoreOptions | TypeCompareKind.CompareNullableModifiersForReferenceTypes | TypeCompareKind.UnknownNullableModifierMatchesAny :
-                TypeCompareKind.AllIgnoreOptions;
+            var compareKind = TypeCompareKind.AllIgnoreOptions |
+                (includeNullability ?
+                    TypeCompareKind.CompareNullableModifiersForReferenceTypes |
+                        (unknownNullableModifierMatchesAny ? TypeCompareKind.UnknownNullableModifierMatchesAny : 0) :
+                    0);
             return type1.Equals(type2, compareKind);
         }
 
         private bool HasIdentityConversionInternal(TypeSymbol type1, TypeSymbol type2)
         {
-            return HasIdentityConversionInternal(type1, type2, IncludeNullability);
+            return HasIdentityConversionInternal(type1, type2, IncludeNullability, UnknownNullableModifierMatchesAny);
         }
 
-        internal static bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
+        internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations source, TypeSymbolWithAnnotations destination)
```

Comment:
>HasTopLevelNullabilityIdentityConversion [](start = 22, length = 40)

I'm wondering why the identity conversion case considers `UnknownNullableModifierMatchesAny`. 
Is it required? Could `UnknownNullableModifierMatchesAny` only be considered in implicit conversion method below?

Reply:
Good catch. Fixed.

---
In reply to: [221333891](https://github.com/dotnet/roslyn/pull/29727#discussion_r221333891) [](ancestors = 221333891)